build_runner/lib/src/package_graph/dependency_ordering.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection';
+import 'dart:math' show min;
+
+/// Finds the strongly connected components of an ordered graph using Tarjan's
+/// algorithm.
+///
+/// [V] is the type of values in the graph vertices. [K] must be a type suitable
+/// for using as a Map or Set key, and [key] must provide a consistent key for
+/// ever vertex. [successors] should return the children for a given vertex.
+///
+/// [nodes] must contain at least a root of every tree in the graph if there are
```

Comment:
Consider renaming `nodes` to `rootNodes` or something that indicates better what is required.

Or, just make this take a single root node. Does that not work for all our use cases?

Reply:
Was attempting to match the wikipedia article as close as possible. Same reason I used `successors` below.

I think it doesn't hurt to still take multiple roots, and the algorithm is tolerant to taking all nodes in the graph which is the general way it's used. Our specific case let's us get away with only passing one root.

Reply:
I don't think matching the wiki article is important, especially in terms of variable names. For instance it uses single letter names and other things which don't match dart readability guidelines. I would do what makes the most sense in our context (Dart, package:build).

Fwiw, I also really dislike the `lowLinks` name :).

Reply:
I was also considering whether we should create a `package:graphs` to include this, and maybe `asyncCrawl` from `build_runner`. I looked at the implementation of this algorithm in `pub` and I think it'd be pretty easy to drop in this method.
I like having these types of complex-but-generally-useful things partitioned off into separate packages rather than re-implement them everywhere depending on certain details of the use case.

@kevmoo - any thoughts on whether a new package would be worth the effort? So far 2 use cases for this function, and another function that would fit nicely in the package. It would mean a new dependency in the SDK but could be worthwhile.

Reply:
+1 for putting this in a separate package (pub uses this for dartdevc as well, not sure if that's what you were referring to)

Reply:
are you asking for a 'graph' package?

Reply:
Yes, probably a 'graphs' package since 'graph' is already taken (but dead?)

I'll write up a 1-pager and share it with you internally.

Reply:
Hmm, there is an implementation of this algorithm already published in package:collection.

https://www.dartdocs.org/documentation/collection/1.14.3/collection/stronglyConnectedComponents.html

It has an API which is slightly less easy to use for this case, since we don't already have exactly a `Map<T, Iterable<T>>`. It wouldn't be quite drop in for pub either, while the signature I chose could be.